### PR TITLE
Crm api#2/create mgmt for product

### PR DESCRIPTION
**Pull request by ymalakhov**
Add functionality for products

### CODE REVIEW

File: src/main/java/com/main/server/controllers/ProductController.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.constraints.NotNull;
+import java.util.List;
+
+@RestController
+@CrossOrigin
+@RequestMapping(path = {"/api/organizations/{orgId}"})
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/controllers/ProductController.java:22]: rename it organizationId

File: src/main/java/com/main/server/model/Product.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    private User createdBy;
+
+    @JsonIgnore
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "deleted_by")
+    private User deletedBy;
+
+    @JsonIgnore
+    @OneToMany(mappedBy = "product")
+    private List<Parameter> customParameters;
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/model/Product.java:79]: u need to set cascade type

File: src/main/java/com/main/server/repository/ProductRepository.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public interface ProductRepository extends CrudRepository<Product, Long> {
+
+    Optional<Product> findByNameAndOrganizationId(String name, Long id);
+    Optional<Product> getById(Long id);
+    List<Product> findAllByOrganizationIdAndDeletedByNull(Long orgId);
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/repository/ProductRepository.java:15]: add pagination

File: src/main/java/com/main/server/repository/ProductRepository.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public interface ProductRepository extends CrudRepository<Product, Long> {
+
+    Optional<Product> findByNameAndOrganizationId(String name, Long id);
+    Optional<Product> getById(Long id);
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/repository/ProductRepository.java:14]: rename to findById, read about difference

File: src/main/java/com/main/server/service/ProductService.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    Product saveProduct(Product product, Long orgId) throws ResourceAlreadyExistException;
+
+    Product getProductById(Long id);
+
+    Product updateProduct(Product product, Long orgId);
+
+    Product deleteProduct(Long id);
+
+    List<Product> getAllProducts(Long orgId);
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/service/ProductService.java:20]: should have args: page and size for pagination

File: src/main/java/com/main/server/service/impl/ProductServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        log.info("Enter getProductById() productId: {}", id);
+        return productRepository.getById(id).orElseThrow(
+                () -> new ResourceNotFoundException(
+                        String.format("Product with such id {%d} is not exist", id))
+        );
+    }
+
+    @Override
+    public Product updateProduct(Product product, Long orgId) {
+        log.info("Enter updateProduct() productName: {}", product.name());
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/service/impl/ProductServiceImpl.java:54]: could you add product at message to see the object fields

File: src/main/java/com/main/server/service/impl/ProductServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                .orElse(null);
+        if (exitedProductWithSameName != null) {
+            throw new ResourceAlreadyExistException(
+                    String.format("Product with name %s already exist in this organization", product.name()));
+        }
+
+        exitedProduct.name(product.name());
+        exitedProduct.description(product.description());
+        exitedProduct.price(product.price());
+        exitedProduct.primeCost(product.primeCost());
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/service/impl/ProductServiceImpl.java:67]: I think there is quantity field

File: src/main/resources/db/migration/V1.0.2__crm_table_add_products.sql [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    PRIMARY KEY (id),
+
+    CONSTRAINT fk_product_id
+        FOREIGN KEY (product_id)
+            REFERENCES products (id),
+
+    CONSTRAINT fk_tag_id
+        FOREIGN KEY (tag_id)
+            REFERENCES tags (id)
+);
```
Comment (DimaGalchenko) [inline src/main/resources/db/migration/V1.0.2__crm_table_add_products.sql:81]: Add line to end

File: src/main/resources/db/migration/V1.0.2__crm_table_add_products.sql [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        FOREIGN KEY (created_by)
+            REFERENCES users (id)
+);
+
+ALTER TABLE custom_parameters
+    ADD CONSTRAINT fk_product_id
+        FOREIGN KEY (product_id)
+            REFERENCES products (id);
+
+CREATE TABLE IF NOT EXISTS tags
```
Comment (DimaGalchenko) [inline src/main/resources/db/migration/V1.0.2__crm_table_add_products.sql:58]: rename to product_tags

File: src/main/java/com/main/server/controllers/ProductController.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+@AllArgsConstructor
+public class ProductController {
+
+    private ProductService productService;
+
+    @PostMapping("/product")
+    public ResponseEntity<Product> saveProduct(@RequestBody Product product, @PathVariable @NotNull Long orgId) {
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/controllers/ProductController.java:29]: Add permission, user must have Creator or Admin organization role

File: src/main/java/com/main/server/controllers/ProductController.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        return ResponseEntity.ok(productService.saveProduct(product, orgId));
+    }
+
+    @GetMapping("/product/{id}")
+    public ResponseEntity<Product> getProductById(@PathVariable @NotNull Long id) {
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/controllers/ProductController.java:34]: Add permission, user must have View permission

File: src/main/java/com/main/server/controllers/ProductController.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        return ResponseEntity.ok(productService.getProductById(id));
+    }
+
+    @PutMapping("/product")
+    public ResponseEntity<Product> updateProduct(@RequestBody Product product, @PathVariable @NotNull Long orgId) {
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/controllers/ProductController.java:39]: Same as save

File: src/main/java/com/main/server/controllers/ProductController.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        return ResponseEntity.ok(productService.updateProduct(product, orgId));
+    }
+
+    @DeleteMapping("/product/{id}")
+    public ResponseEntity<Product> deleteProduct(@PathVariable @NotNull Long id) {
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/controllers/ProductController.java:44]: same as save

File: src/main/java/com/main/server/controllers/ProductController.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        return ResponseEntity.ok(productService.deleteProduct(id));
+    }
+
+    @GetMapping("/products")
+    public ResponseEntity<List<Product>> getAllProducts(@PathVariable @NotNull Long orgId) {
```
Comment (DimaGalchenko) [inline src/main/java/com/main/server/controllers/ProductController.java:49]: Should have View permission

**Review by DimaGalchenko — CHANGES_REQUESTED** [review]
LGTM, minor notes

File: src/main/java/com/main/server/repository/ProductRepository.java (hunk 3 above)
Comment (ymalakhov) [reply src/main/java/com/main/server/repository/ProductRepository.java:15]: I`ll add it in another task

File: src/main/java/com/main/server/service/ProductService.java (hunk 5 above)
Comment (ymalakhov) [reply src/main/java/com/main/server/service/ProductService.java:20]: Another task

**Pull request merged**